### PR TITLE
Rework spinners to track full rotations

### LINKED ISSUE #8224: Spinners can be cheesed
**Describe the bug:** 

By tapping extremely rapidly with two fingers with the circle of the spinner at the middle. You can go up to 430 RPM.

**Screenshots or videos showing encountered issue:** 

[Video](https://player.vimeo.com/video/396941387) 

**osu!lazer version:** 

2020.311.0 

**Logs:**
 Android

### PR DESCRIPTION
**Pull request by smoogipoo**
Resolves https://github.com/ppy/osu/issues/8224

This is a WIP PR for anyone interested to provide comments and play around with it ahead of time. It hasn't been rigorously tested, the test cases are lacking, and it can probably be optimised more (haven't profiled), but the general idea is here.

The general idea is for a spin to require a full 360deg arc to be completed. Rewinding is difficult because "undoing" that arc is non-trivial, so I'm storing the angle at each turning point of the arc such that the change in angle for that segment is guaranteed to be monotonic.

### CODE REVIEW

File: osu.Game.Rulesets.Osu/Skinning/Default/SpinnerRotationTracker.cs [hunk 1]
```diff
@@ -107,9 +113,78 @@ public void AddRotation(float angle)
...
+                }
+
+                currentDirection = direction;
+            }
+            else if (turningPoints.Count > 0 && time <= turningPoints[^1].time)
+            {
+                currentDelta = turningPoints[^1].delta;
+                currentDirection = Math.Sign(currentDelta);
+                turningPoints.RemoveAt(turningPoints.Count - 1);
+            }
```
Comment (peppy) [inline osu.Game.Rulesets.Osu/Skinning/Default/SpinnerRotationTracker.cs:141]: This is going to need some heavy inline commenting.. I'm having difficulty understanding the top level branches, and the importance of the fall-through condition (if that's a valid case).

File: osu.Game.Rulesets.Osu/Skinning/Default/SpinnerRotationTracker.cs [hunk 2]
```diff
@@ -107,9 +113,78 @@ public void AddRotation(float angle)
...
+
+                if (direction != 0 && direction != currentDirection)
+                {
+                    turningPoints.Add((time, currentDelta));
+                    currentDelta = 0;
+                }
+
+                currentDirection = direction;
+            }
+            else if (turningPoints.Count > 0 && time <= turningPoints[^1].time)
```
Comment (peppy) [inline osu.Game.Rulesets.Osu/Skinning/Default/SpinnerRotationTracker.cs:136]: Also assuming this is doing rewind, should it not be looping while `time <= turningPoints[^1].time`?

File: osu.Game.Rulesets.Osu/Skinning/Default/SpinnerRotationTracker.cs [hunk 3]
```diff
@@ -107,9 +114,78 @@ public void AddRotation(float angle)
...
+                // Start a new rotation.
+                // It is fine for totalAfter to skip one rotation (i.e be >= 720), since full rotations are properly accounted for in effectiveRotation.
+                // All that matters is any partial rotation.
+                float excess = (totalAfter % 360) * Math.Sign(delta);
+
+                curAngle = excess;
+                maxAngle = 0;
+                maxAngle = Math.Max(maxAngle, curAngle);
+                minAngle = 0;
+                minAngle = Math.Min(minAngle, curAngle);
```
Comment (peppy) [inline osu.Game.Rulesets.Osu/Skinning/Default/SpinnerRotationTracker.cs:187]: What's going on here? Should this be simplified to 

```suggestion
                maxAngle = Math.Max(0, curAngle);
                minAngle = Math.Min(0, curAngle);
```

File: osu.Game.Rulesets.Osu/Skinning/Default/SpinnerRotationTracker.cs [hunk 4]
```diff
@@ -107,9 +114,78 @@ public void AddRotation(float angle)
...
+            float minAngle = 0;
+            float effectiveRotation = 0;
+
+            for (int i = 0; i < turningPoints.Count; i++)
+                applyDelta(turningPoints[i].delta);
+            applyDelta(currentDelta);
+
+            return effectiveRotation;
+
+            void applyDelta(float delta)
```
Comment (peppy) [inline osu.Game.Rulesets.Osu/Skinning/Default/SpinnerRotationTracker.cs:162]: I know you said this can be optimised, but I wasn't expecting this to run through every point every frame 😅

**Comment by peppy** [comment]
In practice this does seem to work, but I'm having serious trouble parsing the code.

File: osu.Game.Rulesets.Osu/Skinning/Default/SpinnerRotationTracker.cs (hunk 4 above)
Comment (smoogipoo) [reply osu.Game.Rulesets.Osu/Skinning/Default/SpinnerRotationTracker.cs:162]: Yeah I mean there is probably a simpler way to do this that I haven’t figured out yet.

File: osu.Game.Rulesets.Osu/Skinning/Default/SpinnerRotationTracker.cs (hunk 2 above)
Comment (smoogipoo) [reply osu.Game.Rulesets.Osu/Skinning/Default/SpinnerRotationTracker.cs:136]: Fair point

File: osu.Game.Rulesets.Osu/Skinning/Default/SpinnerRotationTracker.cs (hunk 3 above)
Comment (smoogipoo) [reply osu.Game.Rulesets.Osu/Skinning/Default/SpinnerRotationTracker.cs:187]: This is intentional. It resets to 0 and then does the min/max. `curAngle` is an arbitrary value that is either positive or negative. It could be written as:
```
if (curAngle > 0)
{
    maxAngle = curAngle;
    minAngle = 0;
}
else if (curAngle < 0)
{
    maxAngle = 0;
    minAngle = curAngle;
}
else
{
    maxAngle = 0;
    minAngle = 0;
}
```

Comment (smoogipoo) [reply osu.Game.Rulesets.Osu/Skinning/Default/SpinnerRotationTracker.cs:187]: It does not match functionality.

Comment (peppy) [reply osu.Game.Rulesets.Osu/Skinning/Default/SpinnerRotationTracker.cs:187]: Can you double-check? I think there's a misunderstanding somewhere here.

Comment (smoogipoo) [reply osu.Game.Rulesets.Osu/Skinning/Default/SpinnerRotationTracker.cs:187]: You're right, I think I misread something.

File: osu.Game.Rulesets.Osu/Skinning/Default/SpinnerRotationTracker.cs [hunk 5]
```diff
@@ -77,6 +80,10 @@ protected override void Update()
             Rotation = (float)Interpolation.Damp(Rotation, currentRotation, 0.99, Math.Abs(Time.Elapsed));
         }
 
+        private float currentDelta;
+        private int currentDirection;
+        private readonly List<(double time, float delta)> turningPoints = new List<(double, float)>();
```
Comment (bdach) [inline osu.Game.Rulesets.Osu/Skinning/Default/SpinnerRotationTracker.cs:85]: This isn't being cleared in `resetState()` but should be.

Failure to do this manifests on maps with several spinners, where due to pooling and instance reuse, clearing one spinner will cause every subsequent spinner on the map to pretty much auto-clear.

File: osu.Game.Rulesets.Osu/Skinning/Default/SpinnerRotationTracker.cs [hunk 6]
```diff
@@ -107,9 +114,76 @@ public void AddRotation(float angle)
...
+            {
+                currentDelta = turningPoints[^1].delta;
+                currentDirection = Math.Sign(currentDelta);
+                turningPoints.RemoveAt(turningPoints.Count - 1);
+            }
+
+            currentDelta += rateAdjustedRotation;
+            drawableSpinner.Result.RateAdjustedRotation = computeEffectiveRotation();
+
+            Logger.Log(currentDelta.ToString(CultureInfo.InvariantCulture));
```
Comment (bdach) [inline osu.Game.Rulesets.Osu/Skinning/Default/SpinnerRotationTracker.cs:146]: Probably should remove at some point

```suggestion
```

File: osu.Game.Rulesets.Osu/Skinning/Default/SpinnerRotationTracker.cs (hunk 4 above)
Comment (bdach) [reply osu.Game.Rulesets.Osu/Skinning/Default/SpinnerRotationTracker.cs:162]: Well the simplest way would probably be extra memory, wouldn't it? Store (time, momentary effective rotation total) pairs in another list or something, and use the last momentary total to apply the delta onto. On rewind, pop the totals until stack only contains items in the past relative to the rewound position.

Don't have many more smarter ideas than that for now. If we only cared about time instants where spins are stored, we could only store that, but we need the instantaneous effective rotation values too for proper rewind playback.

File: osu.Game.Rulesets.Osu/Skinning/Default/SpinnerRotationTracker.cs (hunk 1 above)
Comment (bdach) [reply osu.Game.Rulesets.Osu/Skinning/Default/SpinnerRotationTracker.cs:141]: I dunno, seemed pretty straightforward to me when I read it - what's unclear here specifically?

File: osu.Game.Rulesets.Osu/Skinning/Default/SpinnerRotationTracker.cs (hunk 5 above)
Comment (peppy) [reply osu.Game.Rulesets.Osu/Skinning/Default/SpinnerRotationTracker.cs:85]: Also makes me wonder if this will even do what it's intending to do when fixed. Wouldn't it break rewinds? Does this data need to be stored somewhere persistent (non-pooled) per spinner?

File: osu.Game.Rulesets.Osu/Skinning/Default/SpinnerRotationTracker.cs (hunk 1 above)
Comment (peppy) [reply osu.Game.Rulesets.Osu/Skinning/Default/SpinnerRotationTracker.cs:139]: Having the `else` condition also include subcondition reduces readability. It's also unclear what anything is doing anywhere. I'd go as far as saying in the current state, with no inline commenting, this reads about on-par with stable's spinner code.

I'll resolve this one and do a more thorough code review pass to add my thoughts.

File: osu.Game.Rulesets.Osu/Skinning/Default/SpinnerRotationTracker.cs [hunk 7]
```diff
@@ -107,9 +112,74 @@ public void AddRotation(float angle)
...
+            float rateAdjustedRotation = (float)(angle * rate);
+
+            if (rate > 0)
+            {
+                int direction = Math.Sign(rateAdjustedRotation);
+
+                if (direction != 0 && direction != currentDirection)
+                {
+                    turningPoints.Add((time, currentDelta));
+                    currentDelta = 0;
```
Comment (peppy) [inline osu.Game.Rulesets.Osu/Skinning/Default/SpinnerRotationTracker.cs:129]: Is there a reason we can't just store the cumulative delta to `turningPoints`? If there is, it's not immediately obvious (so some kind of commenting explaining why on the relevant code would be appreciated).

File: osu.Game.Rulesets.Osu/Skinning/Default/SpinnerRotationTracker.cs [hunk 8]
```diff
@@ -77,6 +78,10 @@ protected override void Update()
             Rotation = (float)Interpolation.Damp(Rotation, currentRotation, 0.99, Math.Abs(Time.Elapsed));
         }
 
+        private float currentDelta;
```
Comment (peppy) [inline osu.Game.Rulesets.Osu/Skinning/Default/SpinnerRotationTracker.cs:81]: Rename to "deltaSinceLastDirectionChange"? If this name seems correct then it may be enough to document what this is. Else xmldoc would be required.

File: osu.Game.Rulesets.Osu/Skinning/Default/SpinnerRotationTracker.cs [hunk 9]
```diff
@@ -107,9 +112,74 @@ public void AddRotation(float angle)
             }
 
             currentRotation += angle;
+
```
Comment (peppy) [inline osu.Game.Rulesets.Osu/Skinning/Default/SpinnerRotationTracker.cs:115]: Unrelated to this PR's changes, but `rotationTransferred` usage above needs explanation because I have no idea what it's doing.

File: osu.Game.Rulesets.Osu/Skinning/Default/SpinnerRotationTracker.cs [hunk 10]
```diff
@@ -107,9 +112,74 @@ public void AddRotation(float angle)
...
+                currentDelta = turningPoints[^1].delta;
+                currentDirection = Math.Sign(currentDelta);
+                turningPoints.RemoveAt(turningPoints.Count - 1);
+            }
+
+            currentDelta += rateAdjustedRotation;
+            drawableSpinner.Result.RateAdjustedRotation = computeEffectiveRotation();
+        }
+
+        private float computeEffectiveRotation()
```
Comment (peppy) [inline osu.Game.Rulesets.Osu/Skinning/Default/SpinnerRotationTracker.cs:145]: This needs xmldoc to explain what "effective" rotation is.

I guess one of the reasons I'm struggling with this code is that we have...

`currentDelta`
`currentRotation`
`effectiveRotation`
`rateAdjustedRotation`
`lastAngle`

with little documentation as to what each one means. gets very hard to follow.

File: osu.Game.Rulesets.Osu/Skinning/Default/SpinnerRotationTracker.cs (hunk 2 above)
Comment (peppy) [reply osu.Game.Rulesets.Osu/Skinning/Default/SpinnerRotationTracker.cs:134]: I'd rewrite this section like this, for better readability:

```diff
diff --git a/osu.Game.Rulesets.Osu/Skinning/Default/SpinnerRotationTracker.cs b/osu.Game.Rulesets.Osu/Skinning/Default/SpinnerRotationTracker.cs
index f580635565..3f6ba3f72e 100644
--- a/osu.Game.Rulesets.Osu/Skinning/Default/SpinnerRotationTracker.cs
+++ b/osu.Game.Rulesets.Osu/Skinning/Default/SpinnerRotationTracker.cs
@@ -119,7 +119,18 @@ public void AddRotation(float angle)
             double rate = gameplayClock?.GetTrueGameplayRate() ?? Clock.Rate;
             float rateAdjustedRotation = (float)(angle * rate);
 
-            if (rate > 0)
+            bool isRewinding = rate > 0;
+
+            if (isRewinding)
+            {
+                while (turningPoints.Count > 0 && time <= turningPoints[^1].time)
+                {
+                    currentDelta = turningPoints[^1].delta;
+                    currentDirection = Math.Sign(currentDelta);
+                    turningPoints.RemoveAt(turningPoints.Count - 1);
+                }
+            }
+            else
             {
                 int direction = Math.Sign(rateAdjustedRotation);
 
@@ -131,12 +142,6 @@ public void AddRotation(float angle)
 
                 currentDirection = direction;
             }
-            else if (turningPoints.Count > 0 && time <= turningPoints[^1].time)
-            {
-                currentDelta = turningPoints[^1].delta;
-                currentDirection = Math.Sign(currentDelta);
-                turningPoints.RemoveAt(turningPoints.Count - 1);
-            }
 
             currentDelta += rateAdjustedRotation;
             drawableSpinner.Result.RateAdjustedRotation = computeEffectiveRotation();

```

File: osu.Game.Rulesets.Osu/Skinning/Default/SpinnerRotationTracker.cs (hunk 5 above)
Comment (bdach) [reply osu.Game.Rulesets.Osu/Skinning/Default/SpinnerRotationTracker.cs:83]: Oh yeah good point. This needs to be stored in the judgement to be 100% correct.

**Comment by peppy** [comment]
We've had a bit of discussion IRL over this (mostly centered around simplifying the implementation). I brought up the point that there's no real reason to store each turning point, instead only tracking the known "zero point" aka the angle that the mouse was pressed at along with subsquent full rotation times (already stored as ticks).

Expressing this on a one dimensional plane looks like it should be pretty simple, but it fails to handle the part where a spinner is a circle, and when rewinding, without replaying all frames since the previous tick, we cannot discern between:

- A user at 340 degrees having spun backwards 20 degrees
- A user at 340 degrees having spun forwards 340 degrees

Which means we still need to store the "direction of spin since the last tick". At which point it's probably approaching the same complexity as the implementation in this PR.

@smoogipoo I think it may be okay to go with what you have here, applying the review changes I proposed in the thread.

As per our discussion, I still think using a relative zero point makes it far easier to parse the code, and to explain to a user, but I worry that altering the implementation to make it work is going to take an equally long amount of time to get right.
